### ISSUE TITLE
How to sample one node in each graph, given the node probabilities?

### ISSUE THREAD

**Issue by jaromiru** [rusty1s/pytorch_geometric#1161]
## ❓ Questions & Help

In the batch of graphs, I first create a probability for each node (via `torch_geometric.utils.softmax`) and then want to sample one node for each graph. If all graphs were of uniform size, obviously the way is with `torch.distributions.Categorical`. However, here I can only think of a sequential way of sampling the nodes one at each time.

Is there a way to do this efficiently in Pytorch-Geometric?

**Comment by rusty1s** [comment]
Yes, that's tricky and we do not have an implementation for this.  The way I can think of it is to either make use of sequential sampling for each graph, or transform your probabilities to uniform size (with p=0 for fake nodes).

**Comment by jaromiru** [comment]
Allright, a pity though. Currently, I cannot think of anything better neither.

For reference and other, there goes a simple sequential sampling:
`def segmented_sample(probs, splits):
    probs_split = torch.split(probs, splits)
    samples = [torch.multinomial(x, 1) for x in probs_split]
    
    return torch.tensor(samples, dtype=torch.int64)`

Minimal working example:
`import torch, torch_scatter

z = torch.rand(10)
splits = [2, 3, 5]
index = torch.tensor([0, 0, 1, 1, 1, 2, 2, 2, 2, 2])

z_probs = torch_scatter.scatter_softmax(z, index)
samples = segmented_sample(z_probs, splits)

print(samples)`
